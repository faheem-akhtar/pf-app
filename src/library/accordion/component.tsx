import { FunctionComponent, useState } from 'react';

import { IconThinChevronDownTemplate } from 'components/icon';
import { domClassMerge } from 'helpers/dom/class-merge';

import styles from './accordion.module.scss';
import { AccordionComponentPropsInterface } from './component-props.interface';

export const AccordionComponent: FunctionComponent<AccordionComponentPropsInterface> = ({
  expanded = false,
  ...props
}) => {
  const [isExpanded, setIsExpanded] = useState(false || expanded);

  return (
    <div className={styles.accordion}>
      <header className={styles.accordion__header} onClick={(): void => setIsExpanded((isExpanded) => !isExpanded)}>
        <p className={styles.accordion__title}>{props.title}</p>
        <IconThinChevronDownTemplate
          class={domClassMerge(styles.accordion__icon, {
            [styles['accordion__icon--expanded']]: isExpanded,
          })}
          clipped
        />
      </header>

      {isExpanded && props.children}
    </div>
  );
};
