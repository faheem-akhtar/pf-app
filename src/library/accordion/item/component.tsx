import { FunctionComponent, useState } from 'react';

import { IconThinChevronDownTemplate } from 'components/icon';
import { domClassMerge } from 'helpers/dom/class-merge';

import styles from './accordion-item.module.scss';
import { AccordionItemComponentPropsInterface } from './component-props.interface';

export const AccordionItemComponent: FunctionComponent<AccordionItemComponentPropsInterface> = ({
  expanded = false,
  children,
  title,
}) => {
  const [isExpanded, setIsExpanded] = useState(false || expanded);

  return (
    <div className={styles.container}>
      <header className={styles.header} onClick={(): void => setIsExpanded((isExpanded) => !isExpanded)}>
        <p className={styles.title}>{title}</p>
        <IconThinChevronDownTemplate
          class={domClassMerge(styles.icon, {
            [styles['icon--expanded']]: isExpanded,
          })}
          clipped
        />
      </header>

      {isExpanded && <div className={styles.panel}>{children}</div>}
    </div>
  );
};
