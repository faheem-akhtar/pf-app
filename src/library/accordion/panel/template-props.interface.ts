export interface AccordionPanelTemplatePropsInterface {
  /**
   * Whether the panel is list item or not
   * @default false
   */
  listItem?: boolean;
}
