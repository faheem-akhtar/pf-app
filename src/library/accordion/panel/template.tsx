import { FunctionComponent } from 'react';

import styles from '../accordion.module.scss';
import { AccordionPanelTemplatePropsInterface } from './template-props.interface';

export const AccordionPanelTemplate: FunctionComponent<AccordionPanelTemplatePropsInterface> = ({
  listItem = false,
  children,
}) => {
  const Component = listItem ? 'li' : 'div';

  return (
    <Component className={styles.accordion__panel} role={listItem ? 'listitem' : undefined}>
      {children}
    </Component>
  );
};
