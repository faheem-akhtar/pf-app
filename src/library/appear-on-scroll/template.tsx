import { domClassMerge } from 'helpers/dom/class-merge';

import { AppearOnScrollStatusEnum } from './status.enum';
import { AppearOnScrollTemplatePropsInterface } from './template-props.interface';

import styles from './appear-on-scroll.module.scss';

const classByStatus = {
  [AppearOnScrollStatusEnum.ENTERING]: styles.appear_on_scroll__entering,
  [AppearOnScrollStatusEnum.EXITING]: styles.appear_on_scroll__exiting,
  [AppearOnScrollStatusEnum.WRAPPER_IN_VIEW]: styles.appear_on_scroll__hidden,
  [AppearOnScrollStatusEnum.WRAPPER_OUTSIDE_VIEW]: styles.appear_on_scroll__visible,
};

export const AppearOnScrollTemplate = ({
  status,
  className,
  children,
}: AppearOnScrollTemplatePropsInterface): JSX.Element => (
  <div
    data-testid='AppearOnScrollTemplate'
    className={domClassMerge(styles.appear_on_scroll, classByStatus[status as AppearOnScrollStatusEnum], className)}
  >
    {children}
  </div>
);
