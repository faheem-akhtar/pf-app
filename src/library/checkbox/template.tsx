import { FunctionComponent } from 'react';

import { domClassMerge } from 'helpers/dom/class-merge';

import styles from './checkbox.module.scss';
import { CheckboxTemplatePropsInterface } from './template-props.interface';

/**
 * Checkbox allows users to select one or more options from the given options.
 *
 * Design: https://www.figma.com/file/gRARY1Vi4W2Ow1vRLw02am/PF_Consumer-Web-Kit?node-id=6%3A1581
 */
export const CheckboxTemplate: FunctionComponent<CheckboxTemplatePropsInterface> = (props): JSX.Element => (
  <div className={domClassMerge(styles.container, props.containerClassName)}>
    <input
      id={props.id}
      name={props.name}
      aria-labelledby={props.name}
      type='checkbox'
      checked={props.checked}
      onChange={props.onChange}
      className={domClassMerge(styles.input, props.inputClassName)}
    />
    <label htmlFor={props.id} className={domClassMerge(styles.label, props.labelClassName)}>
      {props.children}
    </label>
  </div>
);
