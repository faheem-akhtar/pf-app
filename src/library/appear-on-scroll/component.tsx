import { useEffect, useRef, useState } from 'react';

import { ElementService } from 'services/element/service';

import { AppearOnScrollComponentPropsInterface } from './component-props.interface';
import { appearOnScrollObserve } from './observe';
import { appearOnScrollOnStatusChange } from './on-status-change';
import { appearOnScrollOnVisibilityChange } from './on-visibility-change';
import { AppearOnScrollStatusEnum } from './status.enum';
import { AppearOnScrollTemplate } from './template';

export const AppearOnScrollComponent = (props: AppearOnScrollComponentPropsInterface): JSX.Element => {
  const wrapper = useRef<HTMLDivElement>(null);
  const prevStatusRef = useRef<AppearOnScrollStatusEnum | null>(null);
  const [status, setStatus] = useState<AppearOnScrollStatusEnum | null>(null);

  // On mount
  useEffect(() => {
    const element = wrapper.current as HTMLDivElement;

    const wrapperClientRec = ElementService.getBoundingClientRect(element);
    const wrapperInView = wrapperClientRec.y >= 0 && wrapperClientRec.y < document.body.clientHeight;
    // set the initial status on mount
    setStatus(wrapperInView ? AppearOnScrollStatusEnum.WRAPPER_IN_VIEW : AppearOnScrollStatusEnum.WRAPPER_OUTSIDE_VIEW);

    // observe the target element, when it's visibility change - call onVisibilityChange
    return appearOnScrollObserve(element, (visible) => appearOnScrollOnVisibilityChange(visible, setStatus));
  }, []);

  if (prevStatusRef.current !== status) {
    appearOnScrollOnStatusChange(props, status);
  }
  prevStatusRef.current = status;

  return (
    <div ref={wrapper}>
      <AppearOnScrollTemplate status={status} className={props.className}>
        {props.children}
      </AppearOnScrollTemplate>
    </div>
  );
};
