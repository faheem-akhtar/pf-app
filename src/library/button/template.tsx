import { Fragment, FunctionComponent } from 'react';

import { domClassMerge } from 'helpers/dom/class-merge';

import { ButtonIconPositionEnum } from './icon-position.enum';
import { ButtonLinkTemplatePropsInterface } from './link-template-props.interface';
import { ButtonTemplatePropsInterface } from './template-props.interface';
import { LoaderTemplate } from 'components/loader/template';

import styles from './button.module.scss';

// TODO-FE[TPNX-3016] add tests
/**
 * Web UI Kit Button tempalte
 * Figma: https://www.figma.com/file/gRARY1Vi4W2Ow1vRLw02am/PF_Consumer-Web-Kit?node-id=3869%3A2340
 */
export const ButtonTemplate: FunctionComponent<ButtonLinkTemplatePropsInterface | ButtonTemplatePropsInterface> = ({
  icon = {},
  componentType,
  ...props
}) => {
  const { component: Icon, position: iconPosition } = icon;

  const renderLoader = (): JSX.Element => (
    <LoaderTemplate containerCssClass={styles.loader} dotCssClass={styles[`${componentType}--loader`]} />
  );

  const renderIcon = (): JSX.Element | null =>
    Icon ? (
      <Icon clipped={false} class={domClassMerge(styles.icon, styles[`icon--${iconPosition}`], icon.className)} />
    ) : null;

  const renderContent = (): JSX.Element => (
    <Fragment>
      {(iconPosition === ButtonIconPositionEnum.left || iconPosition === ButtonIconPositionEnum.top) && renderIcon()}
      {props.children}
      {iconPosition === ButtonIconPositionEnum.right && renderIcon()}
    </Fragment>
  );

  const Component = (props as ButtonLinkTemplatePropsInterface).href ? 'a' : 'button';

  return (
    <Component
      class={domClassMerge(
        styles.button,
        styles[componentType],
        styles[props.size],
        {
          [styles[`${componentType}--loading`]]: !!props.loading,
          [styles.disabled]: !!props.disabled,
          [styles.loading]: !!props.loading,
          [styles[`${componentType}--disabled`]]: !!props.disabled,
        },
        props.className
      )}
      disabled={props.disabled}
      onClick={!props.loading ? props.onClick : undefined}
      href={(props as ButtonLinkTemplatePropsInterface).href}
      target={(props as ButtonLinkTemplatePropsInterface).target}
    >
      {props.loading ? renderLoader() : renderContent()}
    </Component>
  );
};
