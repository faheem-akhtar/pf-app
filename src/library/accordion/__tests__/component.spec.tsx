import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { AccordionComponent } from '../component';
import { AccordionPanelTemplate } from '../panel/template';

describe('AccordionComponent', () => {
  it('should render without throwing any errors', () => {
    const { container } = render(
      <AccordionComponent title='Popular searches' expanded>
        <AccordionPanelTemplate>content</AccordionPanelTemplate>
      </AccordionComponent>
    );

    expect(container).toMatchSnapshot();
  });

  it('should expanded state works properly', () => {
    const { rerender } = render(
      <AccordionComponent title='Popular searches'>
        <AccordionPanelTemplate>content</AccordionPanelTemplate>
      </AccordionComponent>
    );

    expect(screen.queryByText('content')).not.toBeInTheDocument();

    userEvent.click(screen.getByText('Popular searches'));

    rerender(
      <AccordionComponent title='Popular searches' expanded>
        <AccordionPanelTemplate>content</AccordionPanelTemplate>
      </AccordionComponent>
    );

    expect(screen.getByText('content')).toBeInTheDocument();
  });

  it('should render multiple content', () => {
    const items = ['Apartments for rent', 'Villas for rent', 'Townhouses for rent'];

    render(
      <AccordionComponent title='Popular searches' expanded>
        <ul role='list'>
          {items.map((item, index) => (
            <AccordionPanelTemplate key={index} listItem>
              {item}
            </AccordionPanelTemplate>
          ))}
        </ul>
      </AccordionComponent>
    );

    expect(screen.getAllByRole('listitem')).toHaveLength(3);
  });
});
