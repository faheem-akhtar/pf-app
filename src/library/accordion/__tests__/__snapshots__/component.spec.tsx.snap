// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionComponent should render without throwing any errors 1`] = `
<div>
  <div
    class="accordion"
  >
    <header
      class="accordion__header"
    >
      <p
        class="accordion__title"
      >
        Popular searches
      </p>
      <svg
        class="accordion__icon accordion__icon--expanded"
        viewBox="4.5 8 16 8"
      >
        <path
          clip-rule="evenodd"
          d="M4.64645 8.14645C4.47288 8.32001 4.4536 8.58944 4.58859 8.78431L4.64645 8.85355L11.6464 15.8536C11.82 16.0271 12.0894 16.0464 12.2843 15.9114L12.3536 15.8536L19.3536 8.85355C19.5488 8.65829 19.5488 8.34171 19.3536 8.14645C19.18 7.97288 18.9106 7.9536 18.7157 8.08859L18.6464 8.14645L12 14.793L5.35355 8.14645C5.17999 7.97288 4.91056 7.9536 4.71569 8.08859L4.64645 8.14645Z"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
    </header>
    <div
      class="accordion__panel"
    >
      content
    </div>
  </div>
</div>
`;
